keps/77-dynamically-sized-jobs/README.md

```diff
@@ -0,0 +1,260 @@
+
+### Goals
+
+- Gracefully handle resize operations for Kueue jobs (i.e. update quota usage without suspend, enqueue scale ups)
+- Autoscaling RayCluster works with Kueue since it has an autoscaler and there is high demand (MVP)s
+
+### Non-Goals
+
+- Vertical scaling of workloads – Kueue will only handle resize operations that scale Pods horizontally 
+- Support resize for other Kueue jobs such as Job, JobSet, etc (future)
+- Resizing of RayJobs
+- Partial Preemption
+
+## Proposal
+
+Update the Job framework reconciler and introduce new controllers to orchestrate dynamic resizing of jobs. We are only interested in horizontal scaling of jobs (e.g. scaling more replicas). At a high level, this will be accomplished by:
+- Creating Workload Slice objects that represent incremental scale up of jobs. This gives us per-replica control of workload admission. Workload Slices will be garbage collected and consolidated with their parent workloads after successful admission.
+- Adding default scheduling gates to control the scheduling of new pods based on their admission status.
+- Dynamically adjust quotas in ClusterQueue based on scaling events.
+
+For the MVP, we will only focus on admission of **RayClusters** with autoscaling enabled.
+
+### User Stories
+
+#### Story 1 - RayCluster w/ autoscaling
+
+1. The user creates a RayCluster with `enableInTreeAutoscaling: true`.
+2. Kueue admits the RayCluster based on the requested resources in the head pod and worker pod. 
+3. Given the demand of the job, the Ray Autoscaler adjusts the replicas field as it adds or removes Pods from the cluster.
+4. Kueue will not suspend and requeue the RayCluster, instead it will dynamically update ClusterQueue usage based on the scaling event. 
+
+### Notes/Constraints/Caveats (Optional)
+
+If Kueue needs to preempt the resized RayCluster, it would preempt it as a whole, regardless of whether the RayCluster was previously scaled up.
+
+## Design Details
+
+### Workload Slices
+To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.
+
+The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.
+- Workload Slices will be submitted to the same LocalQueue that's referenced by the top-level RayCluster
+- Workload Slices will be created by Kueue and use identical PodTemplates which is already enforced by Kuberay
+- Workload Slices will beneed to belong to the same resource flavor as the top-level RayCluster that was initially admitted
+
+
+### Creating Workload Slices
```

Comment:
Could you clarify how to keep compatibility when the partial admission occurs during the first admission?

Reply:
There is no partial admission for RayClusters.

However, in general, I don't see a problem. The job can still request a resize to match their original requested total size or bigger.

There are some implementation considerations, such as paying attention to the `.count` fields in the Workload status. But I don't see anything particularly problematic.

Reply:
> There is no partial admission for RayClusters.

Yes, that's right. But I'd like to clarify it for batch/v1 Job integration in the future.

> However, in general, I don't see a problem. The job can still request a resize to match their original requested total size or bigger.
> 
> There are some implementation considerations, such as paying attention to the .count fields in the Workload status. But I don't see anything particularly problematic.

That makes sense generally, but what happens when requests are submitted with scale-in counts less than minCount? I'd like to clarify the minCount semantics in this situation.